18 Vectorise matched glass descriptions

georgerichardson opened the pull request:
closes #18 

---

Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have run `flake8` and addressed any linter erors
- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have rebased onto `dev` (or merged any new changes from `dev`)
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [x] Appropriate information has been added to `README`s
- [ ] I have explained the feature in this PR or (better) in `output/reports/`
- [x] I have requested a code review

georgerichardson committed:
add newline at end

georgerichardson committed:
add newline at end

georgerichardson committed:
fix typo

georgerichardson committed:
fix typo

georgerichardson committed:
change to no gpu and add org_ids to datastore

georgerichardson committed:
change to no gpu and add org_ids to datastore

georgerichardson committed:
apply black

georgerichardson committed:
apply black

georgerichardson committed:
add docker hub as batch container registry

georgerichardson committed:
add docker hub as batch container registry

georgerichardson committed:
update to use gpu queue

georgerichardson committed:
update to use gpu queue

georgerichardson committed:
add torch

georgerichardson committed:
add torch

georgerichardson committed:
update to remove model intialisation from encode function

georgerichardson committed:
update to remove model intialisation from encode function

bishax committed:
Code to get flow to run on batch

TODO: Dockerfile and AWS ECR deploy script

bishax committed:
Code to get flow to run on batch

TODO: Dockerfile and AWS ECR deploy script

georgerichardson committed:
remove transformers approach and add truncate org ids for test mode

georgerichardson committed:
remove transformers approach and add truncate org ids for test mode

georgerichardson committed:
remove transformers utils

georgerichardson committed:
remove transformers utils

georgerichardson committed:
fix org ids by fetching from start step

georgerichardson committed:
fix org ids by fetching from start step

georgerichardson committed:
remove commented import

georgerichardson committed:
remove commented import

georgerichardson committed:
change glass_embeddings to glass_description_embeddings

georgerichardson committed:
change glass_embeddings to glass_description_embeddings

georgerichardson committed:
add vectorise glass descriptions step

bishax commented on industrial_taxonomy/getters/glass_embed.py:
```diff
@@ -0,0 +1,32 @@
+"""Data getter for Glass company description embeddings."""
```
I'd suggest placing this in `glass_house.py` because this is giving embeddings for the merged "glass house" dataset.

bishax commented on industrial_taxonomy/getters/glass_embed.py:
```diff
@@ -0,0 +1,32 @@
+
+
+@lru_cache()
+def get_run():
```
This has been refactored out into a util, see e.g.

https://github.com/nestauk/industrial_taxonomy/blob/a9e9db6a0f2e4a5f50f8ec4c897ed2dc9b971b69/industrial_taxonomy/getters/glass_house.py#L7

https://github.com/nestauk/industrial_taxonomy/blob/a9e9db6a0f2e4a5f50f8ec4c897ed2dc9b971b69/industrial_taxonomy/getters/glass_house.py#L53

bishax commented on industrial_taxonomy/getters/glass_embed.py:
```diff
@@ -0,0 +1,32 @@
+def glass_description_embeddings(
+    run: Optional[Run] = None,
+) -> npt.ArrayLike:
+    """ "Gets embeddings of Glass orgsanisation descriptions."""
```
- Formatting error and a typo
- As numpy does not have good type hints yet, I'd suggest adding more information about the output type and shape

bishax commented on industrial_taxonomy/pipeline/glass_embed/flow.py:
```diff
@@ -0,0 +1,100 @@
+        batch_size = 100_000
+        self.org_description_chunks = chunks(org_descriptions, batch_size)
+
+        self.next(self.embed_descriptions, foreach="org_description_chunks")
```
What is the logic for the branching logic?
This spins up `len(self.org_ids) / batch_size` different GPU instances - are `100,000` embeddings that slow with such a powerful instance?

bishax commented on industrial_taxonomy/pipeline/glass_embed/flow.py:
```diff
@@ -0,0 +1,100 @@
+        memory=61000,
+        cpu=8,
+    )
+    @pip(path="requirements.txt")
```
The following suggestion using the `libraries` argument would be more straightforward

```suggestion
    @pip(libraries={"sentence_transformers": "2.1.0"})
```
Then you could delete `requirements.txt`

bishax commented on industrial_taxonomy/pipeline/glass_embed/utils.py:
```diff
@@ -0,0 +1,12 @@
+    chunked = []
+    for i in range(0, len(l), n):
+        chunked.append(l[i : i + n])
+    return chunked
```
https://toolz.readthedocs.io/en/latest/api.html#toolz.itertoolz.partition has you covered here

bishax commented on industrial_taxonomy/pipeline/README.md:
````diff
@@ -36,3 +36,12 @@ python industrial_taxonomy/pipeline/glass_description_ngrams/nlp_flow.py \
+
+```bash
+python industrial_taxonomy/pipeline/glass_embed/flow.py \
+ --environment=conda \
````
I don't think the conda line is necessary here which will shave off some overhead.

(If `embed_descriptions` was running locally it would be worth having it in, so that the `@pip` install of `sentence_transformers` didn't pollute your main environment.)

bishax replied:
Note: Requires `--no-pylint` to run if `--environment=conda` is dropped

bishax commented on industrial_taxonomy/pipeline/glass_embed/flow.py:
```diff
@@ -0,0 +1,100 @@
+
+@project(name="industrial_taxonomy")
+class GlassEmbed(FlowSpec):
+    """Transform descriptions of fuzzy matched companies into embeddings.
```
Any info about the model used (rationale/link to model docs etc.) should go here

bishax commented on industrial_taxonomy/getters/glass_embed.py:
```diff
@@ -0,0 +1,32 @@
+def org_ids(run: Optional[Run] = None) -> List[int]:
+    """Gets IDs of embedded Glass organisations."""
+    run = run or get_run()
+    step = Step(f"GlassEmbed/{run.id}/start")
```
We should get artifacts from `Steps`, artifacts consumed outside of a flow should be available from the `Run().data` object.

Your join step needs `self.merge_artifacts(inputs, include=["org_ids"])` to make the artifact available, see https://docs.metaflow.org/metaflow/basics#data-flow-through-the-graph

bishax commented on industrial_taxonomy/pipeline/glass_embed/flow.py:
```diff
@@ -0,0 +1,100 @@
+
+        self.next(self.embed_descriptions, foreach="org_description_chunks")
+
+    @batch(
+        queue="job-queue-GPU-nesta-metaflow",
+        image="metaflow-pytorch",
+        # Queue gives p3.2xlarge, with:
+        gpu=1,
+        memory=61000,
+        cpu=8,
+    )
```
What speedup factor does the GPU give? I'm curious

georgerichardson replied:
Annecdotally from playing around on and off GPUs with sentence transformers it's pretty big but I haven't actually measured. Something else to check. Takes ~12 mins for that step to run though in production mode.

bishax commented on industrial_taxonomy/pipeline/glass_embed/flow.py:
```diff
@@ -0,0 +1,100 @@
+        self.next(self.join)
+
+    @step
+    def join(self, inputs):
```
Doc-string.

bishax commented on industrial_taxonomy/pipeline/glass_embed/flow.py:
```diff
@@ -0,0 +1,100 @@
+        self.next(self.end)
+
+    @step
+    def end(self):
```
Doc-string.

Even if it's just `"""No-op."""` if means that anyone looking at the flow but not the source code - e.g. `python flow.py show` or in the new Metaflow GUI (that I'm working on deploying) - can get an idea of what the steps do.

georgerichardson replied:
Done. 

Side note: I originally tried to use the `end` step to `del` unwanted artifacts but that doesn't seem to work. Do previous artifacts not get passed down to the last step automatically? (see new version which doesn't use a foreach so merging shouldn't be an issue)

bishax replied:
For linear steps they propagate but in a join step the values of artifacts can be ambiguous (each foreach branch may set the value of an artifact to be different to others) so you have use `self.merge_artifacts(...)` or manually set the artifacts.
See: https://docs.metaflow.org/metaflow/basics#data-flow-through-the-graph

Do you have a minimal example of this not working in a flow?

bishax commented on industrial_taxonomy/pipeline/glass_embed/flow.py:
```diff
@@ -0,0 +1,100 @@
+        from industrial_taxonomy.getters.glass import get_organisation_description
+        from industrial_taxonomy.getters.glass_house import glass_companies_house_lookup
+
+        org_descriptions = get_organisation_description()
+        self.org_ids = org_descriptions.index.intersection(
+            list(glass_companies_house_lookup().keys())
+        ).to_list()
+
+        org_descriptions = org_descriptions.loc[self.org_ids]
+
+        if self.test_mode and not current.is_production:
+            test_size = 200_000
+            org_descriptions = org_descriptions.head(test_size)
+            self.org_ids = self.org_ids[:test_size]
+
+        org_descriptions = org_descriptions["description"].to_list()
```
Not requesting change but flagging that I think I did it "better" here: https://github.com/nestauk/industrial_taxonomy/blob/a9e9db6a0f2e4a5f50f8ec4c897ed2dc9b971b69/industrial_taxonomy/pipeline/glass_description_ngrams/nlp_flow.py#L88-L97 (though your variable naming is better)

I'd argue my version is more concise, flows in a more linear fashion involving less variable mutation, and makes fewer intermediate lists/copies of things

georgerichardson replied:
Agree. I didn't like that bit. Have remodelled it off this

bishax replied:
Short note:

One of the benefits of the declarative below is that we take `nrows` descriptions once on the dataframe and then decompose into the keys and values which means we always end up with them being the same length and don't need to remember to also  filter `org_descriptions` to the new `org_ids`.
https://github.com/nestauk/industrial_taxonomy/blob/a9e9db6a0f2e4a5f50f8ec4c897ed2dc9b971b69/industrial_taxonomy/pipeline/glass_description_ngrams/nlp_flow.py#L88-L97

bishax reviewed (commented):
Runs in test mode and getters (from your non-test run) return 443285 embeddings with 384 dimensions. Have you done any sanity checks / exploration on these?

georgerichardson committed:
Add QA for glass description vectorisation

georgerichardson committed:
move getters to glass_house

georgerichardson committed:
remove glass_embed getters

georgerichardson committed:
remove glass_embed getters

georgerichardson committed:
remove conda env and add no-pylint

georgerichardson committed:
remove conda env and add no-pylint

georgerichardson committed:
Change flow to single batch machine instead of splitting flow

- Flow does not split and produced all embeddings on one GPU batch machine
- Removed requirements.txt and specified dependency in flow decorator
- Custom chunk function replaced by toolz.itertoolz.partition
- Other changes to docstrings and code

georgerichardson committed:
Change flow to single batch machine instead of splitting flow

- Flow does not split and produced all embeddings on one GPU batch machine
- Removed requirements.txt and specified dependency in flow decorator
- Custom chunk function replaced by toolz.itertoolz.partition
- Other changes to docstrings and code

georgerichardson commented:
I haven't run any sanity checks on the results besides checking shape and that order is retained. Will do a manual check of the results before we merge. Small but not insignificant chance another model might work noticeably better

bishax commented on industrial_taxonomy/pipeline/glass_embed/flow.py:
```diff
@@ -4,24 +4,35 @@
     """Transform descriptions of fuzzy matched companies into embeddings.
 
+    This uses the multi-qa-MiniLM-L6-cos-v1 transformer model which encodes up
+    to 512 tokens per document and produced embeddings with 384 dimensions.
```
Does it drop any tokens after 512? How many Glass descriptions does this affect?

georgerichardson replied:
This is now covered by `flow_qa.py`

bishax commented on industrial_taxonomy/pipeline/glass_embed/flow.py:
```diff
@@ -4,24 +4,35 @@
 
+    This uses the multi-qa-MiniLM-L6-cos-v1 transformer model which encodes up
+    to 512 tokens per document and produced embeddings with 384 dimensions.
+    It was trained by chosen as a compromise between speed/size and performance
```
Missing a word in here

georgerichardson replied:
Changed.

n.b this has been updated as I tested a larger model that has higher performance on semantic textual similarity tasks and have stuck with it

bishax commented on industrial_taxonomy/getters/glass_house.py:
```diff
@@ -59,3 +60,23 @@ def get_description_tokens(run: Optional[Run] = None) -> Dict[str, List[str]]:
+
+    Returns a 2d array of size (n, m) where n is the number of companies and
+    m is the length of each embedding.
+    """
```
> industrial_taxonomy/getters/glass_house.py:67:1: DAR201 Missing "Returns" in Docstring: - return
industrial_taxonomy/getters/glass_house.py:67:1: DAR101 Missing parameter(s) in Docstring: - run

bishax commented on industrial_taxonomy/pipeline/glass_embed/flow.py:
```diff
@@ -4,24 +4,35 @@
 import numpy as np
```
> industrial_taxonomy/pipeline/glass_embed/flow.py:4:1: F401 'numpy as np' imported but unused

bishax commented on industrial_taxonomy/pipeline/glass_embed/flow.py:
```diff
@@ -4,24 +4,35 @@
 import numpy.typing as npt
 
+# from toolz.itertoolz import partition
+
```
```suggestion
```

bishax commented on industrial_taxonomy/pipeline/glass_embed/flow.py:
```diff
@@ -4,24 +4,35 @@
     pass
 
-from industrial_taxonomy.pipeline.glass_embed.utils import chunks
+# from industrial_taxonomy.pipeline.glass_embed.utils import chunks
```
```suggestion
```

bishax commented on industrial_taxonomy/getters/glass_house.py:
```diff
@@ -59,3 +60,23 @@ def get_description_tokens(run: Optional[Run] = None) -> Dict[str, List[str]]:
+    """Gets IDs of embedded Glass organisations."""
+    # run = run or get_run()
+    run = run or get_run("GlassEmbed")
+    # step = Step(f"GlassEmbed/{run.id}/start")
```
```suggestion
    run = run or get_run("GlassEmbed")
```

bishax commented on industrial_taxonomy/pipeline/README.md:
```diff
@@ -41,7 +41,7 @@ python industrial_taxonomy/pipeline/glass_description_ngrams/nlp_flow.py \
- --environment=conda \
  --production \
+ --no-pylint \
  run
```
I'd recommend running the whole flow with batch otherwise the upload of the org descriptions artifact in the start step (over home internet) to the S3 datastore is horribly slow.

```suggestion
 run \
 --with batch
```

bishax commented on industrial_taxonomy/pipeline/glass_embed/flow.py:
```diff
@@ -48,19 +61,14 @@ def start(self):
+        self.org_ids = self.org_ids[:nrows]
 
-        org_descriptions = org_descriptions["description"].to_list()
+        self.org_descriptions = org_descriptions.loc[self.org_ids][
```
:lady_beetle:  This isn't restricted to `nrows` !

One of the benefits of the below is that we take `nrows` descriptions once on the dataframe and then decompose into the keys and values which means we always end up with them being the same length.
https://github.com/nestauk/industrial_taxonomy/blob/a9e9db6a0f2e4a5f50f8ec4c897ed2dc9b971b69/industrial_taxonomy/pipeline/glass_description_ngrams/nlp_flow.py#L88-L97

bishax reviewed (commented):
A methodological question, and a few bits of house-keeping then I think we're good to go.

georgerichardson committed:
fix docstrings and add getters for encoder name and org descriptions

georgerichardson committed:
fix docstrings and add getters for encoder name and org descriptions

georgerichardson committed:
remove redundant imports and fix docstring

georgerichardson committed:
remove redundant imports and fix docstring

georgerichardson committed:
add qa flow and description length check

georgerichardson committed:
add qa flow and description length check

georgerichardson committed:
change qa util imports to uncommented

georgerichardson committed:
change qa util imports to uncommented

georgerichardson committed:
add qa getters for glass embeddings

georgerichardson committed:
add qa getters for glass embeddings

georgerichardson committed:
change to all-distillroberta-v1 model

georgerichardson committed:
change to all-distillroberta-v1 model

georgerichardson committed:
add util for random selection

georgerichardson committed:
add util for random selection

georgerichardson committed:
add nearest neighbour matching for glass description embeddings

georgerichardson committed:
add nearest neighbour matching for glass description embeddings

georgerichardson committed:
remove deletion of org_descriptions attribute

georgerichardson committed:
remove deletion of org_descriptions attribute

georgerichardson commented:
There is now a quality assurance flow `flow_qa.py` which produces a chart and number that show the percent of descriptions that are not truncated by the maximum input length of the sentence encoder. It also produces a random sample of 100 company descriptions that are matched to their nearest neighbour by cosine distance according to the embeddings produced.

bishax commented on industrial_taxonomy/getters/glass_house.py:
```diff
@@ -67,16 +67,49 @@ def description_embeddings(
+    return run.data.model_name
+
+
+def qa_tokenized_length_hist_fig(run: Optional[Run] = None) -> str:
+    """Cumulative histogram of tokenized Glass description lengths."""
+    run = run or get_run("GlassEmbedQA")
+    return run.data.tokenized_length_hist_fig
+
+
+def qa_percent_not_truncated(run: Optional[Run] = None) -> str:
+    """Percent of descriptions not truncated by the model max input length."""
+    run = run or get_run("GlassEmbedQA")
+    return run.data.percent_not_truncated
+
+
+def qa_embedding_matches(run: Optional[Run] = None) -> str:
+    """Pairs of Glass organisations matched by the cosine similarity of
+    their embeddings."""
+    run = run or get_run("GlassEmbedQA")
+    return run.data.embedding_matches
```
A question I don't have an answer to... should there be getters for QA? (if so) should getters for QA go in a separate folder?

Thoughts?

georgerichardson replied:
Getter pro:
- QA elements can be investigated individually depending on user's interest

Getter cons:
- Anyone inspecting from outside the project has to install the repo and run the code
- A user has to fire up iPython/Jupyter/Python shell, find the getters they're interested in and run them every time there is a new QA update

As I was working on this, my thinking was that a preferable solution lo-fi for QA outputs would be a markdown file with the charts and statistics. An interactive dashboard might be the way to go in some cases (e.g. OJO style projects), but there would be a lot of overhead to setting up a custom interface for each project and including the option to inspect a variety of data types (e.g. here, we have text outputs in a specific format).

bishax replied:
Once I get secure authentication sorted for the new Metaflow UI then I think the best default approach for presenting QA is likely to be the new Metaflow cards feature:
https://outerbounds.com/blog/integrating-pythonic-visual-reports-into-ml-pipelines/
https://docs.metaflow.org/metaflow/visualizing-results

georgerichardson committed:
change embedding matching to compare across models used

georgerichardson committed:
change embedding matching to compare across models used

georgerichardson committed:
remove conda_base import

georgerichardson committed:
remove conda_base import

georgerichardson committed:
change model to sentence-transformers/paraphrase-distillroberta-v2

georgerichardson committed:
change model to sentence-transformers/paraphrase-distillroberta-v2

georgerichardson committed:
fix embedding match dict to update for each model

georgerichardson committed:
fix embedding match dict to update for each model

bishax commented on industrial_taxonomy/getters/glass_house.py:
```diff
@@ -87,10 +87,29 @@ def embedded_org_ids(run: Optional[Run] = None) -> List[int]:
+    return run.data.model_name
+
+
+def qa_tokenized_length_hist_fig(run: Optional[Run] = None) -> str:
+    """Cumulative histogram of tokenized Glass description lengths."""
+    run = run or get_run("GlassEmbedQA")
+    return run.data.tokenized_length_hist_fig
+
+
+def qa_percent_not_truncated(run: Optional[Run] = None) -> str:
+    """Percent of descriptions not truncated by the model max input length."""
+    run = run or get_run("GlassEmbedQA")
+    return run.data.percent_not_truncated
+
+
+def qa_embedding_matches(run: Optional[Run] = None) -> str:
+    """Pairs of Glass organisations matched by the cosine similarity of
+    their embeddings."""
+    run = run or get_run("GlassEmbedQA")
+    return run.data.embedding_matches
```
Output type annotations are listed as `str`

bishax commented on industrial_taxonomy/pipeline/glass_embed/flow_qa.py:
```diff
@@ -0,0 +1,132 @@
+        descriptions along with their nearest neighbours found using
+        their embeddings.
+        """
+        import faiss
+
+        org_ids = embedded_org_ids()[: self.nrows]
+        descriptions = embedded_org_descriptions()[: self.nrows]
+        sample_idx = random_sample_idx(org_ids, 100)
+
+        model_runs = get_latest_runs_by_model()
+
+        self.embedding_matches = {
+            org_ids[sid]: {"sample_glass_description": descriptions[sid]}
+            for sid in sample_idx
+        }
+        for run in model_runs:
+
+            embeddings = description_embeddings(run)[: self.nrows]
+            model_name = encoder_name(run)
+
+            dimensions = embeddings.shape[1]
+            index = faiss.IndexFlatIP(dimensions)
+            index.add(embeddings)
+
+            distances, nearest_neighbours_idx = index.search(embeddings[sample_idx], 2)
+            distances = distances[:, 1]  # nearest neighbours that aren't self
+            nearest_neighbours_idx = nearest_neighbours_idx[:, 1]
+
+            for sample_id, nearest_id, dist in zip(
+                sample_idx, nearest_neighbours_idx, distances
+            ):
+                org_id = org_ids[sample_id]
+                self.embedding_matches[org_id][model_name] = {
+                    "nearest_glass_org_id": org_ids[nearest_id],
+                    "nearest_glass_description": descriptions[nearest_id],
+                    "distance": dist,
+                }
```
This is a little unwieldy

georgerichardson replied:
Have made this a little more concise, but would prefer to merge than spend much more time on it unless you have any tips!

bishax replied:
Agree, no point in spending longer on it but e.g. the first step would be breaking things up into smaller pieces, e.g. along the lines of

```python
...

def find_nearest_neighbours_idx(*args):
    return "blah"

def a_function(run, *args):
    nearest_neighbours_idx = find_nearest_neighbours_idx(...)
    for sample_id, nearest_id, dist in zip(...):
        org_id = "blah"
        self.embedding_matches[org_id][model_name] = "blah"

# step body from L101
for run in model_runs:
    a_function(run,  ...)
```

bishax replied:
(I was looking at the original not the updated because githubs UI seems to be having a mare)

bishax commented on industrial_taxonomy/pipeline/glass_embed/qa_utils.py:
```diff
@@ -0,0 +1,81 @@
+
+
+@lru_cache
+def get_latest_runs_by_model() -> List[Run]:
+    """Gets the last successful production run for each encoder.
+
+    Returns:
+        latest_runs: A list of the last successful production run for each
+            embedding model used.
+    """
+    runs = Flow("GlassEmbed").runs("project_branch:prod")
+
+    _model_names = []
+    latest_runs = []
+    for run in filter(lambda run: run.successful, runs):
+        try:
+            model_name = run.data.model_name
+        except KeyError:
+            continue
+
+        if model_name not in _model_names:
+            latest_runs.append(run)
+            _model_names.append(model_name)
+
+    return latest_runs
```
For your awareness: Metaflow has a mechanism for managing different branches, e.g. prototyping improvements to models https://docs.metaflow.org/going-to-production-with-metaflow/coordinating-larger-metaflow-projects#custom-branches

But nothing wrong with what you've done here

bishax reviewed (commented):
Having the QA flow is super useful!

(Note: Only cast a brief eye over the code for the QA)

Commit:
glass clustering flow + util scripts (#27)

* glass clustering flow + util scripts

* updated scripts and readmes

* changes in scripts

* changed assign_cluster parametre

* updated readme

* address PR review comments

* add topsbm script

* changed doc clustering flow description

* fixed type error in get_sector, moved glass_sic lookup to glass_house getters, put reverse_dict in new utils script

* removed pip clone of topsbm from requirements

* final scripts

* added merge artifacts to store models

* addressing PR review changes

* latest round of reviews

* fixed type for one of the parameters

* fixed typing error in assigned_shares

* make tokenised descriptions an argument in sector corpora

* added text_sectors script (#32)

* added text_sectors script

* addressed pr review comments

georgerichardson committed:
fix type hints

georgerichardson committed:
fix type hints

georgerichardson committed:
update simplify embedding_neighbours step

georgerichardson committed:
update simplify embedding_neighbours step

georgerichardson committed:
add function to get nearest neighbour to an embedding

georgerichardson committed:
add function to get nearest neighbour to an embedding

georgerichardson committed:
resolve merge conflict in step running order

georgerichardson committed:
Merge branch '18_vectorise_glass' of github.com:nestauk/industrial_taxonomy into 18_vectorise_glass
